Allow saving and publishing posts

nylen opened the pull request:
This PR adds the capability to publish new posts and update existing posts via the WP REST API.  Closes #526.

This PR is the first time we actually need to store all of the data for a post object in the Redux state.  It looks like our `combineUndoableReducers` utility expects to have all undoable actions reside within a single key in the state tree, and the post properties should be part of the undo history.  Accordingly, I've renamed `state.blocks` to `state.editor`:

- `state.blocks.byUid` &rarr; `state.editor.blocksByUid`
- `state.blocks.order` &rarr; `state.editor.blockOrder`
- newly introduced `state.editor.post`

Even though we're not using Backbone in this project, I'm using the [`wp-api.js` Backbone client](https://developer.wordpress.org/rest-api/using-the-rest-api/backbone-javascript-client/) to save posts because the library is already included in WP core and is the simplest way to make an API call.  The alternative would be to use jQuery or another AJAX library, which would involve some error-prone bits of setup that are already handled for us (like building the API URLs and making sure they work with all the different permalink structures).

Currently the state of the publish/update operation is reflected in the main action button text.  I expect we'll want to change this later, but there are 8 possible messages currently:

- **Publish** (editing a new post; no action has been taken)
- **Publishing...** (in the process of publishing a new post; the button is disabled)
- **Published!** (just published a new post)
- **Publish failed** (an error occurred)
- **Update** (editing an existing post; no action has been taken)
- **Updating...** (in the process of updating an existing post; the button is disabled)
- **Updated!** (just updated an existing post)
- **Update failed** (an error occurred)

Until the plugin is more stable, I wonder if we should update the default action to be creating a new draft instead of immediately publishing a post?

nylen committed:
Add post object to editor state

nylen committed:
Add basic publish button component

nylen committed:
Do not use `wp_localize_script` for post content

It treats values as strings instead of performing proper JSON encoding:

https://core.trac.wordpress.org/browser/tags/4.7.4/src/wp-includes/class.wp-scripts.php?marks=443,447#L423

nylen committed:
Implement saving/publishing posts

Including Redux state for tracking network requests/errors/etc.

nylen committed:
Update page URL after creating a new post

nylen committed:
Make eslint happy again

nylen committed:
Make the test suite happy again

joyously commented:
> but there are 8 possible messages currently

Does this include what the Contributor role would see? That role can only save as draft, not publish. Don't the messages come from the server?

nylen commented:
That's a good point about some users not being able to publish.  I think this is another reason to limit the plugin to create new drafts or update existing posts for now - keep in mind that we haven't built the controls to change post status yet.

youknowriad commented on editor/header/tools/publish-button.js:
```diff
@@ -0,0 +1,125 @@
+		buttonText = wp.i18n.__( 'Update' );
+	} else {
+		buttonText = wp.i18n.__( 'Publish' );
+	}
```
I know some people don't like it but I prefer to use `switch (true)` in these cases.

```js
const buttonEnabled = ! isRequesting;
switch( true ) {
  case isRequesting && requestIsNewPost:
    buttonText = wp.i18n.__( 'Publishing...' );
    break;
  case isRequesting && !requestIsNewPost:
    buttonText = wp.i18n.__( 'Publishing...' );
    break;

   // ...
}
```

IMO it's more readable.

nylen replied:
I'd probably change this to a single `if` chain without nesting instead, like `if ( isRequesting && requestIsNewPost )`.  I'll try it

youknowriad commented on editor/header/tools/publish-button.js:
```diff
@@ -0,0 +1,125 @@
+			post,
+			isNew,
+		} );
+	} );
```
Is it time to start separating the action creators into their own `actions.js` file. This would ease testing these actions. I'm also thinking about testing the `mergeWithPrevious` action this way.

nylen replied:
Yes, this definitely doesn't belong here.

youknowriad commented on editor/state.js:
```diff
@@ -204,17 +223,56 @@ export function isSidebarOpened( state = false, action ) {
+ * @param  {Object} action Dispatched action
+ * @return {string}        Updated state
+ */
+export function api( state = {}, action ) {
```
Should this be renamed `saving` or something like that. Thinking we may have other unrelated requests in the future.

youknowriad reviewed (commented):
Great work @nylen, This is testing well!

Should we drop the "Published" or "Updated" message in favor of "Update" when the post is made dirty again?

nylen commented:
> Should we drop the "Published" or "Updated" message in favor of "Update" when the post is made dirty again?

Yes - do we have the mechanism in place yet to detect dirty content?

youknowriad commented:
> Yes - do we have the mechanism in place yet to detect dirty content?

We don't for now, we could do these actions: 'INSERT_BLOCK', 'UPDATE_BLOCK', 'SWITCH_BLOCK_TYPE', 'MOVE_BLOCK_DOWN', 'MOVE_BLOCK_UP' and 'REMOVE_BLOCK'. Granted this is bot very convenient.

nylen commented:
Ok - this should also include typing (content changes), so I'd prefer to wait until this is implemented fully and we have an `isDirty` state we can easily hook into.  So I think what's there is ok for this PR.  I still want to make the following changes today:

- Save drafts instead of publishing posts
- Clean up the messages
- Split `savePost` into a new `actions.js` file and add a couple of tests
- Rename `api` state to `saving` for now; maybe we'll need a better / more flexible mechanism to store request state later on

nylen committed:
Make message determination more readable

nylen committed:
Save drafts instead of publishing posts

nylen committed:
Move `savePost` to a new `actions.js` file

nylen committed:
Rename `state.api` to `state.saving`

nylen commented:
I think this is ready for another round of review.  The latest commits address all feedback except the "dirty state" which I've started to explore in #610.

youknowriad commented on editor/header/tools/publish-button.js:
```diff
@@ -0,0 +1,85 @@
+	( dispatch ) => ( {
+		onUpdate( post, blocks ) {
+			post.content.raw = wp.blocks.serialize( blocks );
+			savePost( dispatch, post );
```
I wonder if we should introduce, `redux-thunk` or any alternative async middleware to simplify this. Could be done later.

nylen replied:
Discussed further at https://github.com/WordPress/gutenberg/pull/594#discussion_r114681675.  As noted there, I'd prefer to explore this separately.

youknowriad reviewed (approved):
Thanks @nylen. This is a great addition towards allowing people to test the plugin with real posts. 👍

mtias commented:
@jasmussen mind giving a quick look at the design implications for the different states?

jasmussen commented:
Really nice work. This is exciting! 

UI wise we should try some changes, so the publish button doesn't do it all. Though I don't think these changes specifically have to happen in _this_ PR —  it works pretty well and is so exciting we shouldn't let a few UI tweaks prevent it from being merged. 

Here are some blueprint mockups. First, a publish dropdown that does not need to happen immediately, if at all. I will open a separate ticket for it, but wanted to show it here regardless:

![publish flow draft](https://cloud.githubusercontent.com/assets/1204802/25656455/59d21804-2ff9-11e7-83f8-d53e0150718b.png)

Publish flow:

![saving publishing](https://cloud.githubusercontent.com/assets/1204802/25656492/902682a0-2ff9-11e7-8698-aaa3f6be1443.png)

Drafting flow:

![drafting](https://cloud.githubusercontent.com/assets/1204802/25656495/94386f5c-2ff9-11e7-969b-e9f6039a18f6.png)

If you agree the above makes sense, I'll open a separate ticket for splitting drafting/saving behavior out into a separate ticket.

nylen commented:
Let's get this one merged as soon as possible and continue design improvements in separate PRs, it's renaming one of the main state keys so there are already a lot of conflicts.

I'll clean those up later today and plan to merge unless anyone has further feedback, in particular @aduth I don't think you've taken a look through here yet.

aduth commented on editor/actions.js:
```diff
@@ -0,0 +1,38 @@
+ */
+import { get } from 'lodash';
+
+export function savePost( dispatch, post ) {
```
Already noted by @youknowriad that we'll need to consider adopting a pattern for asynchronous data flow. This file is perhaps a bit misleading since at a glance I'd expect it to include a set of action creators, but this isn't a true action creator in that it accepts `dispatch` and doesn't return an action object.

```
type ActionCreator = (...args: any) => Action | AsyncAction
```

http://redux.js.org/docs/Glossary.html#action-creator

I'm fine to flesh this out separately; `redux-thunk` is a popular option, but I've become less a fan of it over time, since thunks are hard to test, less predictable, and with the default implementation, encourage access to state via the second `getState` argument (thus breaking the unidirectional data flow).

If I were to suggest an alternative, I think it shouldn't be much trouble to devise a pattern to encourage treating actions with asynchronous effects as _expressing intent_. Using middleware(s), those intents could be mapped to side effects which themselves may incur additional actions.

Roughly something like: https://gist.github.com/aduth/b36ac70f0d3cd3f61faa0b2b81b98a2a

Not set in stone, specifically:

- Couldn't `savePost` just return `REQUEST` ? Maybe we don't need a separate effects mapping. But sometimes the save intent isn't always just about firing a request, but affecting reducer values (e.g. `SAVE_POST` provides semantic meaning that of just `REQUEST` which should cause the UI to update to reflect in-progress save)
- Instead of an effect's `success` property, we could enforce that the request completion itself should be an action expressing intent `REQUEST_COMPLETED` that has its own effect(s) to cause `receivePost` to be dispatched
- Maybe we forgo the `REQUEST` middleware and treat effects as mini-middlewares which have access to `dispatch` and could just use fetch or wp-api.js in their implementations

Anyways, this has been a bit of a brain dump that'll get lost in pull request history, but jotting down now that it's becoming relevant since this has been on my mind lately.

A few additional resources from side projects where I've been applying these ideas:

- https://github.com/aduth/dones/blob/master/src/state/requests/middleware.js
- https://github.com/aduth/dones/blob/master/src/state/dones/effects.js
- https://github.com/aduth/refx

nylen replied:
I would prefer to address this separately as well.  Unlike other parts of this PR (how we actually store the post + its edits in state), it should be pretty easy to refactor this later without touching lots of other parts of the code.

aduth commented on editor/actions.js:
```diff
@@ -0,0 +1,38 @@
+			window.history.replaceState(
+				{},
+				'Post ' + newPost.id,
+				window.location.href.replace( /&post_id=[^&]*$/, '' )
```
Don't know how much we care to address this here, but these are very specific assumptions about URLs where the editor would be loaded. Also breaks down on some valid URLs:

```js
'/admin.php?page=gutenberg&post_id=5'.replace( /&post_id=[^&]*$/, '' )
// Correct: "/admin.php?page=gutenberg"

'/admin.php?post_id=5&page=gutenberg'.replace( /&post_id=[^&]*$/, '' )
// Incorrect: "/admin.php?post_id=5&page=gutenberg"

'/admin.php?page=gutenberg&post_id=5&foo=1'.replace( /&post_id=[^&]*$/, '' )
// Incorrect: "/admin.php?page=gutenberg&post_id=5&foo=1"
```

In a final solution we'll likely want to consider a querystring / URL parser & serializer, such as:

- https://nodejs.org/api/querystring.html
- https://nodejs.org/api/url.html
- https://www.npmjs.com/package/querystringify

nylen replied:
This seems easy enough to address here.

nylen commented on editor/state.js:
```diff
@@ -10,17 +10,35 @@ import { keyBy, last, omit, without } from 'lodash';
+	blocksByUid( state = {}, action ) {
 		switch ( action.type ) {
-			case 'REPLACE_BLOCKS':
+			case 'EDIT_POST':
```
This is a good point (and the above regarding `undoable`), this is probably not stored correctly.  I think we'll end up needing to maintain something like `editor.post.edits` (containing any changed keys, the next one other than `content` will likely be `status`).

Then, separately, do we need to store the full post object **outside** of the undoable reducer to represent what we **think** is the current state of the post on the server, and update it when we receive better information?  Maybe we don't need this functionality yet.

nylen replied:
Update - this PR is currently using the latest known version of the post object to determine if the currently edited post is new (has an ID) or not.  Also lots of interesting uses for this later:  with regular updates from the server, and conflict resolution where appropriate (a huge task, to be fair), we basically have collaborative editing.

nylen commented on editor/state.js:
```diff
@@ -10,17 +10,35 @@ import { keyBy, last, omit, without } from 'lodash';
+				return action.post || state;
+
+			case 'POST_UPDATE_REQUEST_SUCCESS':
+				return action.post;
```
Good point.  Probably the entire post object shouldn't be included in the undoable state, rather just the list of edits like we do in Calypso.  Seems closely related to the thread below as well (https://github.com/WordPress/gutenberg/pull/594#discussion_r114677900).

nylen committed:
Change POST_UPDATE_REQUEST to REQUEST_POST_UPDATE

https://github.com/WordPress/gutenberg/pull/594#discussion_r114675626

nylen committed:
Parse the query string rather than using a regex replace

https://github.com/WordPress/gutenberg/pull/594#discussion_r114683059

nylen committed:
Store post edits in undoable state and full post object in another key

Addresses these review threads:

https://github.com/WordPress/gutenberg/pull/594#discussion_r114678204
https://github.com/WordPress/gutenberg/pull/594#discussion_r114677900
https://github.com/WordPress/gutenberg/pull/594#discussion_r114678428

nylen commented on editor/state.js:
```diff
@@ -208,17 +227,56 @@ export function isSidebarOpened( state = false, action ) {
+export function saving( state = {}, action ) {
+	switch ( action.type ) {
+		case 'POST_UPDATE_REQUEST':
+			return {
```
I tried this out, and it ended up being longer and (IMO) harder to read:

```js
export const saving = combineReducers( {
	requesting: ( state = false, action ) => {
		switch ( action.type ) {
			case 'REQUEST_POST_UPDATE':
				return true;
			case 'REQUEST_POST_UPDATE_SUCCESS':
			case 'REQUEST_POST_UPDATE_FAILURE':
				return false;
		}
		return state;
	},
	successful: ( state = false, action ) => {
		switch ( action.type ) {
			case 'REQUEST_POST_UPDATE_SUCCESS':
				return true;
			case 'REQUEST_POST_UPDATE_FAILURE':
				return false;
		}
		return state;
	},
	error: ( state = false, action ) => {
		switch ( action.type ) {
			case 'REQUEST_POST_UPDATE_SUCCESS':
				return false;
			case 'REQUEST_POST_UPDATE_FAILURE':
				return true;
		}
		return state;
	},
	isNew: ( state = false, action ) => {
		switch ( action.type ) {
			case 'REQUEST_POST_UPDATE':
			case 'REQUEST_POST_UPDATE_SUCCESS':
			case 'REQUEST_POST_UPDATE_FAILURE':
				return action.isNew;
		}
		return state;
	},
} );
```

In this case, at least, I prefer the explicitness of "here are 3 action types and here is how each property changes in response".  I think updating state references and possible re-renders when a network request is fired or completed is OK.

nylen committed:
Fix a typo in the `saving` reducer

nylen committed:
Simplify assignment of `buttonEnabled` flag

https://github.com/WordPress/gutenberg/pull/594#discussion_r114678825

nylen committed:
Use ellipsis character instead of three dots

https://github.com/WordPress/gutenberg/pull/594#discussion_r114679213

nylen commented:
Rebased after #484 was merged and addressed review comments so far.  Behavior is still the same, which I still think is OK for this initial PR, but the latest code notably improves the state shape to avoid a few future problems:

- https://github.com/WordPress/gutenberg/pull/594#discussion_r114678204 - "seems like we'll be able to revert state back to convincing the UI that the post is a new draft, even if we know it's been saved"
- https://github.com/WordPress/gutenberg/pull/594#discussion_r114677900 - "Won't this fall out of sync with `state.editor.post` when receiving the saved post via `POST_UPDATE_REQUEST_SUCCESS` if the content was manipulated on the server?" (collaborative editing!)
- https://github.com/WordPress/gutenberg/pull/594#discussion_r114678428 - "mutating state directly"

aduth commented on editor/actions.js:
```diff
@@ -0,0 +1,43 @@
+			const urlPieces = window.location.href.split( '?' );
+			const qs = parse( urlPieces[ 1 ] || '' );
+			const newUrl = urlPieces[ 0 ] + '?' + stringify( {
+				...qs,
```
Minor: The object spread syntax creates another clone step; this is the equivalent of:

```js
Object.assign( {}, qs, { post_id: newPost.id } );
```

Since we've created and have no plans to reuse the object outside the scope of the function, I think it'd be fine enough to mutate `qs` by assigning directly into it:

```js
Object.assign( qs, { post_id: newPost.id } );
```

nylen replied:
I tried this, and I think the code clarity / elegance due to using the spread operator is worth the extra clone step.  I don't think the performance is a concern here as this is not called very often and the browser's `replaceState` is going to be orders of magnitude more expensive than the `assign`.

aduth commented on editor/state.js:
```diff
@@ -136,6 +156,26 @@ export const blocks = combineUndoableReducers( {
+ * @param  {Object} action Dispatched action
+ * @return {Object}        Updated state
+ */
+export function currentPost( state = {}, action ) {
```
Per your previous comment, still unclear the value or complications we'll encounter with storing the full post object, vs. just the immediate need of `currentPostId` (a [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) consideration).

nylen replied:
Shouldn't most of the UI around current status, title, author, categories, tags, etc. be based on this `currentPost` state key?

aduth commented on editor/state.js:
```diff
@@ -34,7 +40,20 @@ export const blocks = combineUndoableReducers( {
+
+	edits( state = {}, action ) {
+		switch ( action.type ) {
+			case 'EDIT_POST':
```
When do we call `EDIT_POST` ? Or is the idea that we'll use it in follow-up pull requests? How does this relate to `UPDATE_BLOCK` (insert, remove)?

nylen replied:
The idea is that we'll use it in follow-up pull requests, definitely for non-content properties like status, author, taxonomies, ...

Still unclear to me: whether/when to insert the post content into the `edits` state key.  For example, I can imagine something like the Calypso editor where the content is requested from the editor 200ms after the user stops typing.  This seems like potentially a lot of overhead from frequent content serialization though.

aduth commented on editor/test/state.js:
```diff
@@ -9,7 +9,7 @@ import deepFreeze from 'deep-freeze';
  * Internal dependencies
```
Do we want unit tests for `editor.edits`, `saving`, and `currentPost` reducer behaviors?

nylen replied:
Yes.

nylen committed:
Clarify window.location splitting logic using array destructuring

nylen committed:
Add tests for `editor.edits` reducer

nylen committed:
Add tests for `currentPost` reducer

nylen committed:
Add tests for `saving` reducer

nylen committed:
Update dirty state key

nylen commented:
I've addressed or at least responded to all review feedback.  @aduth is there anything else you think we should do in this PR?

aduth reviewed (approved):
I'm going to create a handful of follow-up task issues regarding some of the points we've discussed here, but this should serve as a good base for moving forward 👍
